routing/routing_session.cpp

```diff
@@ -242,6 +242,11 @@ void RoutingSession::Reset()
   m_passedDistanceOnRouteMeters = 0.0;
   m_isFollowing = false;
   m_lastCompletionPercent = 0;
+
+  // reset announcement counters
+  m_routingRebuildCount = -1;
+  m_routingRebuildAnnounceCount = 0;
```

Comment:
Why are these counters needed? Are they really needed? Can you please describe the logic here and in a code comment?

Reply:
routingRebuildCount is an existing counter so I reused that logic to prevent duplicate or unneeded announcements especially for beginning a new route. Notifications are generated as a separate routine from deciding recalculation is necessary, so it would be major refactoring to try and put a notification on the stack directly from the "we need to / are currently recalculating" function. Instead the need for a recalculation announcement is detected when the counters are off (some class state needs to be persisted in between the unrelated functions.)

Reply:
It doesn't look like the right solution. There should be a direct place in the code where recalculation is triggered. Counting on side counters (that can be removed/refactored at any time) is not a robust design.

Reply:
@biodranik I'm not sure if there is. Notifications are generated opportunistically when a state requiring them has been met, and GenerateNotifications can be called any number of times per actual notification message output; the function needs to know *when* it should actually output stuff (like the Boolean for Route IsValid so we don't announce invalid routes.) So we need to track if now is a good time to announce route beginning and route recalculation, and the answer is it's a good time (a) if we haven't yet announced the beginning, state stored separately, and (b) if we've recalculated elsewhere (state stored separately) but not yet announced it (again state somehow persisted across multiple announcements but not between routing sessions)

I'm ok with storing state differently somehow but this was the simplest way I could find

Reply:
I assume that there should be two explicit places in the code, where:
1. Navigation is started (the route is already calculated/present at that moment, so directions should be known/easily calculable).
2. Route rebuild is triggered (this happens when the current location is compared with the existing route).

These two places would be the best to announce.

Does it make sense to split this PR into two parts?
1. Announce route recalculation
2. Announce the first starting direction (only if necessary! e.g. "turn left in 100 meters", not "turn left in 10km").

Reply:
I'm ok with changing "follow the highlighted" to an initial instruction. This has already been merged into yesterday's beta so I'm not sure the overall plan but I'm happy to make future improvements too.

You're correct that there are places when routing is started and recalculation is triggered, but those are not places where *notifications* are triggered. Notifications are placed on the stack via a separate function which is called from a different place so I'm not sure we can directly generate notifications from the route begin or route recalculate functions.

Consider that notifications is a pointer passed into this function and modified. Generating notifications outside of this function would be rearchitecting.

```
GenerateNotifications(std::vector<std::string> & notifications){
  //...
  notifications.emplace_back(m_turnNotificationsMgr.GenerateRecalculatingText());
  return;
}
```

Reply:
@zyphlar Pease don't consider the beta branch as _merged_. It's a temporary branch where we cherry-pick some open PRs that need testing.

@vng did you check that code? As an architect, do you see a better way to trigger these notifications?